src/net.cpp

```diff
@@ -2636,22 +2636,22 @@ void CConnman::StopNodes()
     }
 
     // Close sockets
-    LOCK(cs_vNodes);
-    for (CNode* pnode : vNodes)
+    std::vector<CNode*> nodes;
+    WITH_LOCK(cs_vNodes, nodes.swap(vNodes));
+    for (CNode* pnode : nodes)
         pnode->CloseSocketDisconnect();
     for (ListenSocket& hListenSocket : vhListenSocket)
         if (hListenSocket.socket != INVALID_SOCKET)
             if (!CloseSocket(hListenSocket.socket))
                 LogPrintf("CloseSocket(hListenSocket) failed with error %s\n", NetworkErrorString(WSAGetLastError()));
 
     // clean up some globals (to help leak detection)
-    for (CNode* pnode : vNodes) {
+    for (CNode* pnode : nodes) {
         DeleteNode(pnode);
     }
     for (CNode* pnode : vNodesDisconnected) {
         DeleteNode(pnode);
     }
-    vNodes.clear();
     vNodesDisconnected.clear();
     vhListenSocket.clear();
     semOutbound.reset();
```

Comment:
Might be worth adding a comment as to why `vNodesDisconnected`, `vhListenSocket`, `semOutbound` and `semAddnode` are all safe to reset here -- I think it's because:

 * `vNodesDisconnected` and `vhListenSocket` are only otherwise accessed by `ThreadSocketHandler`
 * `semOutbound` and `semAddnode` are only otherwise accessed via `ThreadOpenConnections`, `Start()`, `Interrupt()` (and `Interrupt()` is only safe if it's invoked in between `Start()` and `StopNodes()`), and `AddConnection` (which is called from rpc/net.cpp so also requires that we won't be adding connections via RPC -- might be good to add an `if (semOutbound == nullptr) return false;` to `AddConnection()`)

Reply:
RPC must be (and is assumed to be) shutdown before `Stop` is called. The other threads are also assumed to be shut down. Maybe reverting my commit (https://github.com/bitcoin/bitcoin/pull/21563#issuecomment-826278304) could help to self-document that better?